### ISSUE TITLE
Update Feature Comparison Table

### ISSUE THREAD

**Issue by artemirq** [shellscape/webpack-plugin-serve#104]
Documentation Is:

- [x] Missing
- [ ] Needed
- [ ] Confusing
- [ ] Not Sure?

### Please Explain in Detail...

It's would be great, if you add comparison with other development server stack:
`Fastify` (better express) + `webpack-hot-middleware` + `webpack-dev-middleware`

For example, I use this stack in my project, [Zero Scripts](https://github.com/zero-scripts/zero-scripts). At [`preset.webpack-spa`](https://github.com/zero-scripts/zero-scripts/blob/evolution/core/packages/preset.webpack-spa/src/WebpackPresetSpa.ts#L27)

### Your Proposal for Changes

Add `Fastify + webpack-hot-middleware + webpack-dev-middleware` to comparison table

**Comment by PlayMa256** [comment]
Personally i don't think it is correct to compare with that stack, here is why. `WPS` is a package that encapsulates all those features into one single package and all the other packages on that table behave the same. It would be fair to compare if we had a solution with all those 3 components into one package!

**Comment by artemirq** [comment]
@PlayMa256 What difference does it make whether it encapsulates these features / packages or not? Proposed variant is one of the best options on replacement `webpack-dev-server`. It would be nice to compare them

**Comment by shellscape** [comment]
@artemirq what he means is that it's difficult to compare a stack versus a pre-packaged product. Both `webpack-dev-server` and `webpack-serve` use `webpack-dev-middleware`. `webpack-dev-server` uses webpack's internal HMR scripts, and `webpack-serve` uses `webpack-hot-client`. Using `webpack-hot-middleware` is providing similar HMR functionality, just without WebSockets or SockJS. Using `fastify` means you're rolling with a different middleware stack than this plugin's `builtins`.

So given all of that, I hope you can see how that's not a good fit for the feature comparison table. It's a lot easier for a grid/table to compare top-level features of a packaged product. It'd turn it into kind of a mess otherwise :smile:

That being said, adding a plain text section to the features doc, or maybe even the FAQ, (about the differences between a similar stack and using WPS out of the box) might be useful to some folks, but that'd require the reader to have a pretty solid working knowledge of servers, middleware, and the differences between the HMR implementations.

**Comment by PlayMa256** [comment]
shellscape summarized what i was really trying to say hahaha!

I really think that this comparison is good to be done, but would fit well on a middleware (with the same principles as  wps) rather than directly here!

**Comment by shellscape** [comment]
If you would like any help with Zero Scripts using this plugin, please let me know. I'm happy to help.

**Comment by PlayMa256** [comment]
We are here to help on anything you need!

**Comment by artemirq** [comment]
Hey @shellscape @PlayMa256. It's available to use `webpack-plugin-serve` without emitting files?
Development mode should not emit files, because this files not used anywhere

```
{
  output: {
    path: !isDev ? paths.build : undefined
  }
}
```
[source](https://github.com/zero-scripts/zero-scripts/blob/master/packages/config.webpack/src/createWebpackConfiguration.ts#L22)

**Comment by PlayMa256** [comment]
We chose to not use `memory-fs` for numerous reasons, most of them performance wise. Also memory-fs is what gives people headache in most problems nowadays (hmr, update, etc.)

**Comment by artemirq** [comment]
@PlayMa256 How to enable HMR, without using `react-hot-loader`?

**Comment by artemirq** [comment]
@PlayMa256 Theoretically `memory-fs` is a storage at RAM, because program running at RAM (indirectly). RAM many times faster than usual file systems, which works with storage disk. Certainly be SSD, but RAM currently faster SSD too. Or i'm wrong?

**Comment by PlayMa256** [comment]
@artemirq afaik no. Actually there is one way, but I have to test before i say something wrong. I'm not closer to a computer now (answering from phone). As I get one I'll test and let you know.

About memory-fs, theoretically yes, it would be faster than disk, but it is not playing nicely with webpack. There are dozens of issues on the webpack repo where the problem is memory-fs, unfortunately, memory overflow, memory leak, all sorts of problems.

Those examples were working, I'm going to check them as I get home and let you know! 😁

**Comment by PlayMa256** [comment]
@artemirq HMR is working on react and on the vue recipes (angular is buggy on my machine, i'm looking at why).

**Comment by artemirq** [comment]
@PlayMa256 Thank you so much. Sorry for wrong information about HMR. Today i found bug at recent Zero Scripts update ([default](https://github.com/zero-scripts/zero-scripts/blob/master/packages/config.webpack/src/createWebpackConfiguration.ts#L60) plugins gonna removed)

**Comment by artemirq** [comment]
@PlayMa256 Can you help with understanding how to enable HMR with `webpack-plugin-serve`? It still doesn't work for me. In previous I thought for lack of `webpack.HotModuleReplacementPlugin`, but it's not true, because `webpack-plugin-serve` automatically adds this plugin

Also see this moments:
- Base Webpack [configuration](https://github.com/zero-scripts/zero-scripts/blob/preset.webpack-spa/change-dev-server/packages/config.webpack/src/createWebpackConfiguration.ts#L18)
- [Instantiate](https://github.com/zero-scripts/zero-scripts/blob/preset.webpack-spa/change-dev-server/packages/preset.webpack-spa/src/extensions/addWebpackPluginServe.ts#L7) and add `webpack-plugin-serve`
- Running development script [start](https://github.com/zero-scripts/zero-scripts/blob/preset.webpack-spa/change-dev-server/packages/preset.webpack-spa/src/WebpackPresetSpa.ts#L17)

You can clone `Zero Scripts` and manually test `React` [example](https://github.com/zero-scripts/zero-scripts/tree/preset.webpack-spa/change-dev-server/examples/react):
1. Clone [Zero Scripts](https://github.com/zero-scripts/zero-scripts)
2. `yarn install`
3. `yarn watch`/`yarn compile`
4. `cd examples/react`
5. `yarn start`

**Comment by shellscape** [comment]
@artemirq have you verified that `watch` is true when you attempt to use HMR?

**Comment by artemirq** [comment]
@shellscape Check screenshot. `wps` client connected with server. And when i edit file, client logs rebuilds (warning from `eslint-loader`). **But changes doesn't applies**
![image](https://user-images.githubusercontent.com/17342159/52182281-a6858e80-280c-11e9-9b6d-bdb420c40c1b.png)

**Comment by shellscape** [comment]
I'm not familiar with React (I know, that's crazy to say these days) but I can tell you from a plain-JS standpoint what I'm _not_ seeing in the log is that HMR was applied, nor do I see a new build notification. You should see a message like this:

![Imgur](https://i.imgur.com/4HRsul5.png)

That tells me there are three possibilities:

1. The WebSocket is not being notified of a new build
1. Webpack isn't watching properly
1. HMR has somehow been disabled

**Comment by shellscape** [comment]
> Maybe warnings prevent apply changes?

Indeed, it looks like that is the case: https://github.com/shellscape/webpack-plugin-serve/blob/0d30966ff133ed73165b603b9f6bb1c74c349f14/lib/routes.js#L50-L64

**Comment by artemirq** [comment]
@shellscape Yeah, when I fixed this and add to `index.js` accepting modules code, HMR is works :)
It is pity that we need all time at every project add this code:

```js
if (module.hot) {
  module.hot.accept();
}
```

Is it possible to automate it by a Webpack Plugin?

**Comment by shellscape** [comment]
According to the documentation, that code is required. https://webpack.js.org/guides/hot-module-replacement/ It may be possible to automate it but I don't know how. 

This issue has gotten pretty far off topic from the original post 😄  If you'd like to open an issue to track allowing HMR with warnings, that would be good. For more discussion on troubleshooting ZeroScripts I think it would be helpful to create a new support issue for that topic.

**Comment by artemirq** [comment]
@shellscape Okay, I create new issue when find a question or bug. And I try to fix it when possible. Thanks for talking. I already created a PR, which fix this (#106)

**Comment by artemirq** [comment]
@shellscape I created [hot-accept-webpack-plugin](https://github.com/artemirq/hot-accept-webpack-plugin), which automatically adds code to accept HMR update. No more repeating of this 😄

```
module.exports = {
  plugins: [
    new HotAcceptPlugin({
      test: /index\.js$/
    })
  ]
};
```

**Comment by shellscape** [comment]
Well done @artemirq!

**Comment by artemirq** [comment]
@PlayMa256 @shellscape Hey again 😀. I'm interested in the topic about `memory-fs` and better solutions for `webpack-plugin-serve`. I don't open new issue, because we should to think about it together. What about this memory-fs [implementation](https://github.com/streamich/memfs)? Maybe this solution doesn't contains problems, which original has? 🤔 ...And we can use `memfs` in `webpack-plugin-serve` 🎉

**Comment by shellscape** [comment]
Hello 😄  Please do open a new issue to discuss this. That topic is very off-topic for this thread.
